src/network/server.rs

```diff
@@ -866,6 +886,573 @@ impl HubService for MyHubService {
+        };
+        let num_shards = self.shard_stores.len();
+        let per_shard_tokens: Vec<Option<Vec<u8>>> = if let Some(token_bytes) = req.page_token {
+            serde_json::from_slice(&token_bytes)
+                .map_err(|e| Status::invalid_argument(format!("Invalid page token: {}", e)))?
+        } else {
+            vec![None; num_shards]
+        };
+        if per_shard_tokens.len() != num_shards {
+            return Err(Status::invalid_argument(
+                "Page token does not match number of shards".to_string(),
+            ));
+        }
+        let pages: Vec<MessagesPage> = self
+            .shard_stores
+            .iter()
+            .zip(per_shard_tokens.into_iter())
+            .map(|(shard_entry, shard_token)| {
+                let page_options = PageOptions {
+                    page_size: req.page_size.map(|s| s as usize),
+                    page_token: shard_token,
+                    reverse: req.reverse.unwrap_or(false),
+                };
+                let cast_store = &shard_entry.1.cast_store;
+                return CastStore::get_casts_by_parent(cast_store, &parent, &page_options)
+                    .unwrap_or(MessagesPage {
+                        messages: vec![],
+                        next_page_token: None,
+                    });
+            })
+            .collect();
+        let combined_messages: Vec<Message> = pages
+            .iter()
+            .flat_map(|page| page.messages.clone())
+            .collect();
+        let next_page_tokens: Vec<Option<Vec<u8>>> =
+            pages.into_iter().map(|page| page.next_page_token).collect();
+        let new_page_token = serde_json::to_vec(&next_page_tokens)
+            .map_err(|e| Status::internal(format!("Failed to serialize next_page_token: {}", e)))?;
+        let response = MessagesResponse {
+            messages: combined_messages,
+            next_page_token: Some(new_page_token),
+        };
+
+        Ok(Response::new(response))
+    }
+
+    async fn get_casts_by_mention(
+        &self,
+        request: Request<FidRequest>,
+    ) -> Result<Response<MessagesResponse>, Status> {
+        let req = request.into_inner();
+        let mention = req.fid;
+
+        let num_shards = self.shard_stores.len();
+
+        let per_shard_tokens: Vec<Option<Vec<u8>>> = if let Some(token_bytes) = req.page_token {
+            serde_json::from_slice(&token_bytes)
+                .map_err(|e| Status::invalid_argument(format!("Invalid page token: {}", e)))?
+        } else {
+            vec![None; num_shards]
+        };
+
+        if per_shard_tokens.len() != num_shards {
+            return Err(Status::invalid_argument(
+                "Page token does not match number of shards".to_string(),
+            ));
+        }
+
+        let pages: Vec<MessagesPage> =
+            self.shard_stores
+                .iter()
+                .zip(per_shard_tokens.into_iter())
```

Comment:
Ah nice, this is a good idea.  I wonder if we should make this explicit by adding a shard id to the request. Probably worth doing sometime after mainnet (and hubs are deprecated).